### PR TITLE
feat(NODE-5241): add option to return modified document

### PR DESCRIPTION
**Pull request by durran**
### Description

Adds a new option to the `findOneAnd...` family of methods, `includeResultMetadata` which defaults to `true`. When set to the default, it will return a `ModifyResult`, when set to `false`, it will return the actual modified document.

#### What is changing?
- Adds the `includeResultMetadata` option.
- Update the types of `findOneAndReplace`, `findOneAndDelete`, `findOneAndUpdate`.
- Adds new type tests.
- Refactors crud api tests and find and modify tests to use async/await.
- Removes redundant tests.

##### Is there new documentation needed for these changes?

None

#### What is the motivation for this change?

NODE-5241

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [x] Changes are covered by tests
- [ ] New TODOs have a related JIRA ticket

### CODE REVIEW

File: src/operations/find_and_modify.ts [hunk 1]
```diff
@@ -29,6 +29,8 @@ export interface FindOneAndDeleteOptions extends CommandOperationOptions {
   sort?: Sort;
   /** Map of parameter names and values that can be accessed using $$var (requires MongoDB 5.0). */
   let?: Document;
+  /** Return the raw result document instead of the ModifyResult */
+  returnRawResult?: boolean;
```
Comment (dariakp) [inline src/operations/find_and_modify.ts:33]: Just wanted to bring a couple of points to your attention from the ticket description so that we minimize potential back and forth in review:

> Dev note: `returnRawResult` is not meant to be prescriptive - the server returns a write_ops::FindAndModifyCommandReply, internally the driver returns a ModifyResult but our flag should make it clearer as to what is being returned

> API docs should clearly explain the option

It may be worth taking this back to the team for consensus on option name; once the option name is settled, it's important to make sure we set accurate expectations for what the option does. The spec compliant version of these commands returns less information than the current `ModifyResult`. In the original ticket, `returnRawResult = true` is intended to return what is currently `ModifyResult` and `returnRawResult = false` is intended to return the spec compliant version, which is equal to `ModifyResult.value`.

File: test/integration/crud/find.test.js [hunk 2]
```diff
@@ -781,126 +781,6 @@ describe('Find', function () {
     }
   });
 
-  /**
-   * Test findOneAndUpdate a document
-   */
-  it('shouldCorrectlyfindOneAndUpdateDocument', {
```
Comment (durran) [inline test/integration/crud/find.test.js:787]: The cases in this giant test case are covered in individual findAndModify tests as well as spec tests.

File: test/integration/crud/find.test.js [hunk 3]
```diff
@@ -981,53 +861,6 @@ describe('Find', function () {
     }
   });
 
-  /**
-   * Test findOneAndUpdate a document
-   */
-  it('Should Correctly Handle findOneAndUpdate Duplicate Key Error', {
```
Comment (durran) [inline test/integration/crud/find.test.js:987]: This was testing server behaviour.

File: src/operations/find_and_modify.ts [hunk 4]
```diff
@@ -127,6 +133,11 @@ class FindAndModifyOperation extends CommandOperation<Document> {
       upsert: false
     };
 
+    // Default to returning the raw result.
+    if (!('includeResultMetadata' in options)) {
+      options.includeResultMetadata = true;
+    }
+
```
Comment (baileympearson) [inline src/operations/find_and_modify.ts:140]:
```suggestion
    options.includeResultMetadata ??= true;
```

File: test/types/schema_helpers.test-d.ts [hunk 5]
```diff
@@ -127,6 +135,41 @@ expectAssignable<SchemaWithIdNumberInterface | null>(
   (await interfaceNumberTestCollection.findOneAndUpdate({ a: 1 }, { a: 5 })).value
 );
 
+expectAssignable<SchemaWithIdNumberType | null>(
```
Comment (baileympearson) [inline test/types/schema_helpers.test-d.ts:138]: Without helpful errors / describe blocks / test names, it's hard to GROK what these tests are testing.  I think they're testing the return type of `findOneAnd*` methods when `includeResultMetadata` is false?

1. should these tests live in findX.test-d.ts?  They're not actually related to how we infer the collection schema (which is that this file is for I think)
2. can we comment them to make it clear what they're testing?  Something I've considered in the past is an anonmyous scope with a comment:
```typescript
// findOneAnd* methods return the modified document when `includeResultMetadata` is false
{ 
	expectAssignable<SchemaWithIdNumberInterface | null>(
  		await interfaceNumberTestCollection.findOneAndDelete({ a: 1 }, { includeResultMetadata: false })
	);
   
	...
} 
```

File: src/operations/find_and_modify.ts (hunk 4 above)
Comment (nbbeeken) [reply src/operations/find_and_modify.ts:140]:
```ts
this.options = { ...options };
```

In either approach, should we be avoiding modifying the input options?

Comment (baileympearson) [reply src/operations/find_and_modify.ts:140]: The options are already copied in `resolveOptions`

File: test/types/schema_helpers.test-d.ts (hunk 5 above)
Comment (durran) [reply test/types/schema_helpers.test-d.ts]: I've reverted the schema_helpers.test-d.ts changes per your comment, and simplified type testing by just adding 3 tests (1 for each method) in findX-test-d.ts and added a comment to them.

File: src/operations/find_and_modify.ts (hunk 4 above)
Comment (durran) [reply src/operations/find_and_modify.ts]: Updated per suggestion.

File: test/integration/crud/find_and_modify.test.ts [hunk 6]
```diff
@@ -0,0 +1,523 @@
...
+        expect(error).to.be.instanceOf(MongoServerError);
+        expect(started).to.have.lengthOf(1);
+        expect(started[0]).to.have.property('query', oid);
+      });
+    });
+
+    context('when passing in writeConcern', {
+      metadata: {
+        requires: { topology: ['single', 'replicaset', 'sharded', 'ssl', 'heap', 'wiredtiger'] }
+      },
```
[the remark this answers is not in the document]
Comment (durran) [reply test/integration/crud/find_and_modify.test.ts]: I've cleaned up all the metadata with the exception of the read preference test which would need at least a replica set.

File: test/integration/crud/find_and_modify.test.ts [hunk 7]
```diff
@@ -0,0 +1,523 @@
...
+    return setupDatabase(this.configuration);
+  });
+
+  describe('#findOneAndDelete', function () {
+    context('when no options are passed', function () {
+      let client;
+      let collection;
+
+      beforeEach(async function () {
+        client = this.configuration.newClient(this.configuration.writeConcernMax(), {
```
[the remark this answers is not in the document]
Comment (durran) [reply test/integration/crud/find_and_modify.test.ts]: I've removed the `writeConcernMax` from the config now, but left `maxPoolSize`. It's not needed, but doesn't harm anything and prevents more connections getting created.

File: test/integration/crud/find_and_modify.test.ts [hunk 8]
```diff
@@ -0,0 +1,523 @@
...
+        await client?.close();
+      });
+
+      it('returns the raw result', async function () {
+        const result = await collection.findOneAndDelete({ a: 1 });
+        expect(result.value.b).to.equal(1);
+      });
+    });
+
+    context('when passing an object id filter', function () {
```
[the remark this answers is not in the document]
Comment (durran) [reply test/integration/crud/find_and_modify.test.ts:33]: I've rephrased the tests per the suggestions, and dryed up the setup/teardown where appropriate.

File: test/integration/crud/find_and_modify.test.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (durran) [reply test/integration/crud/find_and_modify.test.ts]: Yes, I refactored to share where applicable.

File: test/integration/crud/crud_api.test.ts [hunk 9]
```diff
@@ -618,110 +618,173 @@ describe('CRUD API', function () {
...
   });
 
-  it('should correctly execute findAndModify methods using crud api', {
-    // Add a tag that our runner can trigger on
-    // in this case we are setting that node needs to be higher than 0.10.X to run
+  describe('#findOneAndDelete', {
     metadata: {
       requires: { topology: ['single', 'replicaset', 'sharded', 'ssl', 'heap', 'wiredtiger'] }
     },
+    test: async function () {
```
[the remark this answers is not in the document]
Comment (durran) [reply test/integration/crud/crud_api.test.ts]: Removed.

File: test/integration/crud/find_and_modify.test.ts [hunk 10]
```diff
@@ -0,0 +1,393 @@
+import { expect } from 'chai';
+
+import { type CommandStartedEvent, MongoServerError, ObjectId } from '../../mongodb';
+import { setupDatabase } from '../shared';
```
Comment (baileympearson) [inline test/integration/crud/find_and_modify.test.ts:4]: Do we have any tests for `includeResultMetadata = false`?  I see tests for the default behavior, but nothing with `includeResultMetadata` disabled.

File: test/types/community/collection/findX.test-d.ts [hunk 11]
```diff
@@ -305,6 +305,30 @@ expectType<WithId<{ a: number; b: string }> | null>(
   ).value
 );
 
+// Including { includeResultMetadata: false } option will change the return type
```
Comment (baileympearson) [inline test/types/community/collection/findX.test-d.ts:308]: Since we now have specific overloads for all the `findOneAnd*` methods, I think we should add type tests for each.

**Review by baileympearson — CHANGES_REQUESTED** [review]

File: test/types/community/collection/findX.test-d.ts (hunk 11 above)
Comment (durran) [reply test/types/community/collection/findX.test-d.ts:348]: Added tests for all overloads.

File: test/integration/crud/find_and_modify.test.ts (hunk 10 above)
Comment (durran) [reply test/integration/crud/find_and_modify.test.ts:4]: Added tests with the option set to false.

**Review by baileympearson — APPROVED** [review]

File: src/operations/find_and_modify.ts [hunk 12]
```diff
@@ -205,7 +213,7 @@ class FindAndModifyOperation extends CommandOperation<Document> {
     // Execute the command
     super.executeCommand(server, session, cmd, (err, result) => {
       if (err) return callback(err);
-      return callback(undefined, result);
+      return callback(undefined, options.includeResultMetadata ? result : result.value ?? null);
```
Comment (nbbeeken) [inline src/operations/find_and_modify.ts:216]: When is `value` nullish and has it been null or undefined? 

I might be missing them but I'm not seeing a test for the `null` case

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: src/operations/find_and_modify.ts (hunk 12 above)
Comment (durran) [reply src/operations/find_and_modify.ts:216]: Added tests for the non-matching null cases for all 3 methods now.

**Review by baileympearson — APPROVED** [review]
